tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,290 @@
+            gamma,
+            alpha_sp,
+            alpha_stim,
+            coll_ion_coeff,
+            coll_recomb_coeff,
+            partition_function,
+            levels,
+            level_boltzmann_factor,
+        )
+
+        # >>>TODO:initial electron density should be included in the initial guess, added in a future PR
+        initial_electron_density = number_density.sum(axis=0)
+        # <<<
+        rate_matrix = NLTERateEquationSolver.calculate_rate_matrix(
+            phi[0],
+            initial_electron_density[0],
+            rate_matrix_index,
+            photo_ion_rate[0],
+            rad_recomb_rate_coeff[0],
+            coll_ion_coefficient[0],
+            coll_recomb_coefficient[0],
+        )
+        return -1, -1
+
+    @staticmethod
+    def calculate_rate_matrix(
+        phi_shell,
+        electron_density,
+        rate_matrix_index,
+        photo_ion_rate,
+        rad_recomb_rate_coeff,
+        coll_ion_coefficient,
+        coll_recomb_coefficient,
+    ):
+        """
+
+        Parameters
+        ----------
+        phi_shell : DataFrame
+            Saha Factors in the current shell
+        electron_density : float
+            Guess for electron density in the current shell
+        rate_matrix_index : MultiIndex
+            Index used for constructing the rate matrix
+        photo_ion_rate : DataFrame
+            Photo ionization rates
+        rad_recomb_rate_coeff : DataFrame
+            Radiative recombination coefficients(should get multiplied by electron density)
+        coll_ion_coefficient : DataFrame
+            Collisionional ionization coefficients(should get multiplied by electron density)
+        coll_recomb_coefficient : DataFrame
+            Collisionional recombination coefficients(should get multiplied by electron density^2)
+
+        Returns
+        -------
+        DataFrame
+            Rate matrix used for nlte solver.
+        """
+        rate_matrix = pd.DataFrame(
+            0.0, columns=rate_matrix_index, index=rate_matrix_index
+        )
+        rad_recomb_rates = rad_recomb_rate_coeff * electron_density
+        coll_ion_rates = coll_ion_coefficient * electron_density
+        coll_recomb_rates = coll_recomb_coefficient * electron_density**2
+        atomic_numbers = (
+            rate_matrix_index.get_level_values(0).unique()[:-1].values
+        )  # dropping the n_e index
+        for atomic_number in atomic_numbers:
+            ion_numbers = rate_matrix.loc[atomic_number].index.get_level_values(
+                0
+            )
+            phi_block = phi_shell.loc[atomic_number]
+            rate_matrix_block = NLTERateEquationSolver.lte_rate_matrix_block(
+                phi_block, electron_density
+            )
+
+            nlte_ion_numbers = ion_numbers[
+                rate_matrix.loc[atomic_number].index.get_level_values(1)
+                == "nlte_ion"
+            ]
+            lte_ion_numbers = ion_numbers[
```

Comment:
It seem like `lte_ion_numbers` is not used anywhere.

Reply:
I added this initially for the excitation implementation. Should I remove it for now?

Reply:
Done